sql/connect/server/src/main/scala/org/apache/spark/sql/connect/pipelines/PipelineEventSender.scala

```diff
@@ -0,0 +1,185 @@
+  // This will ensure that the sender is properly shut down and all resources are released
+  // without requiring explicit shutdown calls in user code.
+  override def close(): Unit = shutdown()
+
+  /**
+   * Shutdown the event sender, stop taking new events and wait for processing to complete. This
+   * method blocks until all queued events have been processed. Idempotent operation: calling this
+   * multiple times has no effect after the first call.
+   */
+  def shutdown(): Unit = {
+    if (isShutdown.compareAndSet(false, true)) {
+      // Request a shutdown of the executor which waits for all tasks to complete
+      executor.shutdown()
+      // Blocks until all tasks have completed execution after a shutdown request,
+      // disregard the timeout since we want all events to be processed
+      if (!executor.awaitTermination(Long.MaxValue, java.util.concurrent.TimeUnit.MILLISECONDS)) {
+        logError(
+          log"Pipeline event sender for session " +
+            log"${MDC(LogKeys.SESSION_ID, sessionHolder.sessionId)} failed to terminate")
+        executor.shutdownNow()
+      }
+      logInfo(
+        log"Pipeline event sender shutdown completed for session " +
+          log"${MDC(LogKeys.SESSION_ID, sessionHolder.sessionId)}")
+    }
+  }
+
+  /**
+   * Send a single event to the client
+   */
+  private[connect] def sendEventToClient(event: PipelineEvent): Unit = {
+    try {
+      val message = if (event.error.nonEmpty) {
+        // Returns the message associated with a Throwable and all its causes
+        def getExceptionMessages(throwable: Throwable): Seq[String] = {
+          throwable.getMessage +:
+            Option(throwable.getCause).map(getExceptionMessages).getOrElse(Nil)
+        }
+        val errorMessages = getExceptionMessages(event.error.get)
+        s"""${event.message}
+           |Error: ${errorMessages.mkString("\n")}""".stripMargin
+      } else {
+        event.message
+      }
+      responseObserver.onNext(
+        proto.ExecutePlanResponse
+          .newBuilder()
+          .setSessionId(sessionHolder.sessionId)
+          .setServerSideSessionId(sessionHolder.serverSessionId)
+          .setPipelineEventResult(
+            proto.PipelineEventResult.newBuilder
+              .setEvent(
+                proto.PipelineEvent
+                  .newBuilder()
+                  .setTimestamp(
+                    ProtoTimestamp
+                      .newBuilder()
+                      // java.sql.Timestamp normalizes its internal fields: getTime() returns
+                      // the full timestamp in milliseconds, while getNanos() returns the
+                      // fractional seconds (0-999,999,999 ns). This ensures no precision is
+                      // lost or double-counted.
+                      .setSeconds(event.timestamp.getTime / 1000)
+                      .setNanos(event.timestamp.getNanos)
+                      .build())
```

Comment:
nit: let's avoid building so many layers of builder in on function call.

Reply:
@gengliangwang extracted a helper [constructProtoEvent](https://github.com/apache/spark/pull/51956/commits/2bf017c1290470ccf3d94816dc851f07c00b3620) to reduce the nesting